lib/features/build_server/build_provider.dart

```diff
@@ -0,0 +1,265 @@
+
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:flyde/core/networking/protocol/process_completion.dart';
+import 'package:flyde/core/networking/protocol/project_build.dart';
+import 'package:flyde/core/networking/protocol/project_update.dart';
+import 'package:flyde/core/networking/protocol/project_init.dart';
+import 'package:flyde/core/networking/server.dart';
+import 'package:flyde/core/networking/websockets/session.dart';
+import 'package:flyde/features/build_server/cache/cache.dart';
+import 'package:flyde/features/build_server/cache/project_cache.dart';
+import 'package:flyde/features/build_server/compiler_interface.dart';
+
+/// Provides the build service on the local network.
+///
+/// A call to `setup` is mandatory to ensure that connections can be accepted.
+class BuildProvider {
+  /// The server that is used to handle connections.
+  final WebServer _server;
+
+  /// The list of interfaces to isolates which handle compilation.
+  final Map<String, MainInterface> _interfaces = {};
+
+  final Map<String, String> _availableProjects = {};
+
+  /// The cache that is used to store the compiled code.
+  late final Cache _cache;
+
+  /// A queue for each project that manages which client can
+  /// access a project and which have to wait.
+  final Map<String, Queue<ServerSession>> _projectCapacityQueues = {};
+
+  BuildProvider(this._server);
+
+  /// Sets up the server.
+  Future<void> setup({Directory? cacheDirectory}) async {
+    await _server.ready;
+
+    _cache = await Cache.load(from: cacheDirectory);
+    _server.wsOnMessage = _handleWebSocketMessage;
+    _server.wsOnDone = (ServerSession session) {
+      final String? id = session.storage['id'];
+
+      if (id != null) {
+        _activateNextSession(id, removing: session);
+      }
+    };
+  }
+
+  /// Kills all project isolates and shuts down the server.
+  Future<void> terminate() async {
+    _availableProjects.keys.toList().forEach(kill);
+    _availableProjects.clear();
+    _interfaces.clear();
+    await _server.close();
+  }
+
+  /// Terminates the isolate which hosts the project with the id [projectId].
+  void kill(String projectId) {
+    final MainInterface? interface = _interfaces[projectId];
+
+    if (interface != null) {
+      interface.isolate.isolate.kill(priority: Isolate.immediate);
+    }
+
+    if (_availableProjects.containsKey(projectId)) {
+      _availableProjects.remove(projectId);
+    }
+  }
+
+  /// Returns a list of all projects which have an active connection.
+  List<String> get projectIds {
+    return _availableProjects.keys.toList();
+  }
+
+  /// Returns the name of the project with the [projectId].
+  ///
+  /// If the project is not available, an argument error is thrown.
+  String projectName(String projectId) {
+    final String? name = _availableProjects[projectId];
+
+    if (name != null) {
+      return name;
+    }
+
+    throw ArgumentError('Project with id $projectId does not exists');
```

Comment:
Not sure. You could argue that the error is thrown because your argument, the project id, is faulty. This should be prevented by checking if the id is in `activeProjectIds`

Reply:
The argument is only faulty because the internal state does not match. To [quote the docs](https://api.flutter.dev/flutter/dart-core/StateError-class.html):


_The operation was not allowed by the current state of the object._

_Should be used when this particular object is currently in a state which doesn't support the requested operation, but other similar objects might, or the object might change its state to one which supports the operation._